x/cash/dynamicfee.go

```diff
@@ -1,56 +1,58 @@
-package cash
-
-import (
-	"github.com/iov-one/weave"
-	"github.com/iov-one/weave/errors"
-	"github.com/iov-one/weave/gconf"
-	"github.com/iov-one/weave/x"
-)
-
 /*
-DynamicFeeDecorator is an enhanced version the basic FeeDecorator with
-better handling of tx errors, and ability to deduct/enforce
-app-specific fees.
 
-The business logic is:
-
-* If tx fee < min fee, or tx fee cannot be paid, reject with error.
+DynamicFeeDecorator is an enhanced version the basic FeeDecorator with better
+handling of transaction errors and ability to deduct/enforce app-specific fees.
 
-* Run the tx.
-
-* If tx errors, revert all tx changes and charge only the min fee
+The business logic is:
+1. If a transaction fee < min fee, or a transaction fee cannot be paid, reject
+   it with an error.
+2. Run the transaction.
+3. If a transaction processing results in an error, revert all transaction
+   changes and charge only the min fee.
 
-TODO: * If tx succeeded, but requested a RequiredFee higher than paid fee,
-revert all tx changes and refund all but the min fee, returning an error.
+TODO: If a transaction succeeded, but requested a RequiredFee higher than paid
+fee, revert all transaction changes and refund all but the min fee, returning
+an error.
 
-* If tx succeeded, and at least RequiredFee was paid, everything is committed
-and we return success
+If a transaction succeeded, and at least RequiredFee was paid, everything is
+committed and we return success
 
 It also embeds a checkpoint inside, so in the typical application stack:
+
 	cash.NewFeeDecorator(authFn, ctrl),
 	utils.NewSavepoint().OnDeliver(),
+
 can be replaced by
+
 	cash.NewDynamicFeeDecorator(authFn, ctrl),
 
-As with FeeDecorator, all deducted fees are send to the collector,
-whose address is configured via gconf package.
+As with FeeDecorator, all deducted fees are send to the collector, whose
+address is configured via gconf package.
+
 */
+
+package cash
+
+import (
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/gconf"
+	"github.com/iov-one/weave/x"
+)
+
 type DynamicFeeDecorator struct {
-	auth    x.Authenticator
-	control FeeController
-	// these are cached values to not hit gconf on each read
-	collector weave.Address
```

Comment:
no more caching?
i really don't want to hit the db 2-3 times for every request (this decorator is executed everywhere)

Reply:
I was not sure about it. I think this is a performance penalty for being correct.

Avoiding the database lookup is good. I question the consequences. If the decorator will cache those values forever, no dynamic change is possible. 

I think it would be better if the `gconf` package could do the caching. Depending on the requirements it might be an easy task. Today `gconf` does not support messages so only genesis can define configuration. If we want to keep it like this, adding cache in `gconf` itself is easy. If we want to support messages, cache expiration becomes a topic and might require more research.

Regardless of the above, I think that keeping the cache in `gconf` is a better solution. What do you think?

Reply:
I have created an issue for this topic https://github.com/iov-one/weave/issues/361